### PR TITLE
HDX-10119 Add "scan" functionality for analytical and maintenance operations on all of HDX

### PR DESCRIPTION
**Pull request by IanHopkinson**
## Purpose
Version for this PR: 2024.8.2

The aim of this PR is to provide functionality for actions on all datasets in HDX. This uses the `package_search` endpoint in CKAN directly to download the whole catalogue. The actions supported are `survey` which counts occurrences of a key, `distribution` which calculates the distribution of values in a key, `list` which lists the values of a key for each dataset (like the existing list command)  and `delete_key` which deletes a selected key. This final action was the original purpose of the update - to remove `resources._csrf_token` keys, this functionality is limited to only accept the `_csrf_token` and `extras` keys as arguments.  

- [HDX-10119 - Detect and remove the "_csrf_token" at resource level.](https://humanitarian.atlassian.net/browse/HDX-10119)

Under this PR some investigation was done to look at a false report of an "extras" key error. The check for the extras key in parsing the traceback is not specific enough and can mask other errors and the test is not actually testing for the error on the live site.

## Major file changes
This PR adds the `ckan_utilities.py` and `test_ckan_utilities.py` files which implement the scan functionality. The `DEMO.md` file is renamed to `USERGUIDE.md` and updated to include the new command, and reorganised to reflect the more mature status of the project. 

## Minor file changes
`tests/test_hdx_utilities_integration.py` appears to have changed a lot but this is just a result of a change in line endings configuration. The new content in this file is the test function `test_get_hdx_url_and_key`, which tests the new `get_hdx_url_and_key` function.

## Versioning

`hdx-cli-toolkit` uses the CalVer versioning scheme with format YYYY.MM.Micro i.e. 2022.12.1 which is updated manually in `pyproject.toml`. The "Micro" component is simply an integer increased by 1 at each version, starting from 0.
- [x] Version updated in `pyproject.toml` and PR description
- [x] Update README.md and DEMO.md with any new CLI commands

### CODE REVIEW

File: src/hdx_cli_toolkit/ckan_utilities.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
+#!/usr/bin/env python
+# encoding: utf-8
+
+import json
+import urllib3
```
Comment (turnerm) [inline src/hdx_cli_toolkit/ckan_utilities.py:5]: This plus `ckanapi` should be added to the requirements

File: src/hdx_cli_toolkit/ckan_utilities.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    headers = {
+        "Authorization": hdx_api_key,
+        "Content-Type": "application/json",
+    }
+
+    start = 0
+    if "start" not in query.keys():
+        query["start"] = start
+    if "rows" not in query.keys():
+        query["rows"] = 10
```
Comment (turnerm) [inline src/hdx_cli_toolkit/ckan_utilities.py:27]: Where does this 10 come from?

File: src/hdx_cli_toolkit/cli.py [hunk 3]
```diff
@@ -734,3 +744,130 @@ def remove_extras_key(
...
+    t0 = time.time()
+    fetch_all = False
+    if rows is None:
+        start = 0
+        rows = 1000
+        fetch_all = True
+    if fetch_all and input_path is None:
+        print(
+            "No rows value provided so fetching all data in 1000 row chunks. "
+            "This takes ~10 minutes and generates an 865MB file.",
```
Comment (turnerm) [inline src/hdx_cli_toolkit/cli.py:822]: It was only 4 minutes for me!

**Review by turnerm — COMMENTED** [review]
I'm sorry Ian I did not get very far, I can pick it up again once I'm back on the 9th. But feel free to merge in the meantime of course.

File: src/hdx_cli_toolkit/ckan_utilities.py (hunk 1 above)
Comment (IanHopkinson) [reply src/hdx_cli_toolkit/ckan_utilities.py:5]: I've added it for explicitness but I believe it already imported by `hdx-python-api`

File: src/hdx_cli_toolkit/cli.py (hunk 3 above)
Comment (IanHopkinson) [reply src/hdx_cli_toolkit/cli.py:822]: Get you and your fancy internet ;-) Anyway 4 is the same as 10 in physicist order of magnitude terms

File: src/hdx_cli_toolkit/ckan_utilities.py (hunk 2 above)
Comment (IanHopkinson) [reply src/hdx_cli_toolkit/ckan_utilities.py:27]: `Rows` is the `ckanapi` equivalent of `limit` and 10 is the default value I picked / default if queried in the browser. I think I'll put it in a constant and set it to 100.

**Comment by IanHopkinson** [comment]
@turnerm I added a `list` action to `scan` which replicates the `list` command but applies by default to all of HDX. It is more performant than I expected <10s to show the values of 2 keys for all datasets (based on a cached version of the data). I expect it will run in >1s for you ;-)

**Pull request merged**